[SPARK-30039][SQL] CREATE FUNCTION should do multi-catalog resolution

planga82 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Add CreateFunctionStatement and make CREATE FUNCTION go through the same catalog/table resolution framework of v2 commands.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
It's important to make all the commands have the same table resolution behavior, to avoid confusing
CREATE FUNCTION namespace.function

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->
Yes. When running CREATE FUNCTION namespace.function Spark fails the command if the current catalog is set to a v2 catalog.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Unit tests.

AmplabJenkins commented:
Can one of the admins verify this patch?

planga82 commented:
cc: @cloud-fan @viirya @imback82 @dongjoon-hyun

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
ok to test.

SparkQA commented:
**[Test build #115339 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115339/testReport)** for PR 26890 at commit [`23920b2`](https://github.com/apache/spark/commit/23920b2c4e0ae2cc677e303df842bd87e1eb7827).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20145/
Test PASSed.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -475,47 +475,45 @@ class ResolveSessionCatalog(
+  }
+
+  private def parseSessionCatalogFunctionIdentifier(sql: String, catalog: CatalogPlugin,
+      functionIdent: Identifier): (Option[String], String) = {
```
```scala
-  private def parseSessionCatalogFunctionIdentifier(sql: String, catalog: CatalogPlugin,
+  private def parseSessionCatalogFunctionIdentifier(
+      sql: String,
+      catalog: CatalogPlugin,
```

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -475,47 +475,45 @@ class ResolveSessionCatalog(
-          } else {
-            throw new AnalysisException ("SHOW FUNCTIONS is only supported in v1 catalog")
-          }
+          val (db, fn) =
```
Maybe, `val (database, function) =` for consistency.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -475,47 +475,45 @@ class ResolveSessionCatalog(
-      }
-      DescribeFunctionCommand(functionIdentifier, extended)
+      val (database, function) =
+        parseSessionCatalogFunctionIdentifier("DESCRIBE FUNCTION", catalog, functionIdent)
```
Just a question. Is it hard to make `SessionCatalogAndFunction`?

planga82 replied:
It would be possible in the case that we would have v2 catalog functions to resolve it. But for now we need to handle an specific error for this case

`throw new AnalysisException(s"$sql is only supported in v1 catalog")`

If we don't handle it here, we find and unresolved exception that is not clear.
It is the same situation that we have with the commands that use parseV1Table function.

planga82 replied:
I'm not sure what are you proposing exactly. We could have something like that:
  
```
case DescribeFunctionStatement(SessionCatalogAndFunction(catalog, functionIdent), extended) =>
...
case DropFunctionStatement(SessionCatalogAndFunction(catalog, functionIdent), ifExists, isTemp) =>
...
case CreateFunctionStatement(SessionCatalogAndFunction(catalog, functionIdent),
      className, resources, isTemp, ignoreIfExists, replace) =>
...
```
where SessionCatalogAndFunction is an extractor that tests if isSessionCatalog, something like SessionCatalogAndTable do.
In the case that !isSessionCatalog it implies that any of this cases match and the result is 
`unresolved operator 'CreateFunctionStatement [testcat, ns1, ns2, fun]` for Create function 

How could we catch this case generically and throw a generic exception like you propose?

imback82 replied:
Ah OK. This requires throwing an exception inside the extractor, which may not be ideal.

SparkQA commented:
**[Test build #115339 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115339/testReport)** for PR 26890 at commit [`23920b2`](https://github.com/apache/spark/commit/23920b2c4e0ae2cc677e303df842bd87e1eb7827).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateFunctionStatement(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115339/
Test PASSed.

SparkQA commented:
**[Test build #115355 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115355/testReport)** for PR 26890 at commit [`c74974c`](https://github.com/apache/spark/commit/c74974cf487094f7f431050784b11fb17bf5e379).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20162/
Test PASSed.

SparkQA commented:
**[Test build #115355 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115355/testReport)** for PR 26890 at commit [`c74974c`](https://github.com/apache/spark/commit/c74974cf487094f7f431050784b11fb17bf5e379).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115355/
Test PASSed.

planga82 commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -475,47 +475,47 @@ class ResolveSessionCatalog(
+        replace)
+  }
+
+  private def parseSessionCatalogFunctionIdentifier(
```
oh, yes, it sounds good!

SparkQA commented:
**[Test build #115368 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115368/testReport)** for PR 26890 at commit [`74275ff`](https://github.com/apache/spark/commit/74275ffc5225ed4d204ffe43ea50e831baef11b3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20174/
Test PASSed.

imback82 commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -475,47 +475,47 @@ class ResolveSessionCatalog(
+        parseSessionCatalogFunctionIdentifier("DROP FUNCTION", catalog, functionIdent)
+      DropFunctionCommand(database, function, ifExists, isTemp)
+
+    case CreateFunctionStatement(CatalogAndIdentifier(catalog, functionIdent),
```
I would use `ident` when you use `CatalogAndIdentifier` instead of `functionIdent`.

SparkQA commented:
**[Test build #115368 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115368/testReport)** for PR 26890 at commit [`74275ff`](https://github.com/apache/spark/commit/74275ffc5225ed4d204ffe43ea50e831baef11b3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115368/
Test PASSed.

SparkQA commented:
**[Test build #115409 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115409/testReport)** for PR 26890 at commit [`707b037`](https://github.com/apache/spark/commit/707b03704056dbc95a4360aacf269ff780d28201).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20213/
Test PASSed.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -474,48 +474,48 @@ class ResolveSessionCatalog(
+        case Seq(db, fn) => FunctionIdentifier(fn, Some(db))
+        case Seq(fn) => FunctionIdentifier(fn, None)
+        case _ =>
+          throw new AnalysisException(s"Unsupported function name '${functionIdent.quoted}'")
```
This is not caused by this PR. This error message looks too general, and I do not see much info from it.

Maybe we can make it clearer? Like v1 function name should not have multiple namespaces?

planga82 replied:
In the show columns statement we decide this message
> Namespace name should have only one part if specified
How about the same message or similar? Both are clearer

SparkQA commented:
**[Test build #115409 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115409/testReport)** for PR 26890 at commit [`707b037`](https://github.com/apache/spark/commit/707b03704056dbc95a4360aacf269ff780d28201).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115409/
Test PASSed.

planga82 committed:
[SPARK-30039][SQL] CREATE FUNCTION multicatalog resolution


[SPARK-30039][SQL] Pull requests comments


[SPARK-30039][SQL] Pull request comments


[SPARK-30039][SQL] Pull request comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20977/
Test PASSed.

SparkQA commented:
**[Test build #116184 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116184/testReport)** for PR 26890 at commit [`3796a3a`](https://github.com/apache/spark/commit/3796a3a2388c04e3896a57973a133fc481a64578).

SparkQA commented:
**[Test build #116184 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116184/testReport)** for PR 26890 at commit [`3796a3a`](https://github.com/apache/spark/commit/3796a3a2388c04e3896a57973a133fc481a64578).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateFunctionStatement(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116184/
Test FAILed.

planga82 committed:
Fix scala style

SparkQA commented:
**[Test build #116185 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116185/testReport)** for PR 26890 at commit [`d9eb441`](https://github.com/apache/spark/commit/d9eb4410d5c77234746d13132f3cad5aa6092d1d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20978/
Test PASSed.

SparkQA commented:
**[Test build #116185 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116185/testReport)** for PR 26890 at commit [`d9eb441`](https://github.com/apache/spark/commit/d9eb4410d5c77234746d13132f3cad5aa6092d1d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116185/
Test PASSed.

planga82 commented:
Hi @cloud-fan @viirya @imback82 @dongjoon-hyun! Any more comments on this?

cloud-fan commented:
thanks, merging to master!
